### DISCUSSION TITLE
useFieldArray with checkbox input

### DISCUSSION THREAD

**Discussion by GabrielGomesBarbosa** [react-hook-form/react-hook-form#2430] [Q&A] [answered]
Hi @bluebill1049  sorry for asking again.... i trying to render a list with checkbox input using `useFieldArray`, but... when i click on checkbox, didn`t work. Can you provide some example on here or documentation ? (Im using v5)

My Code:

```
const { register, watch, control } = useFormContext();
const { fields } = useFieldArray({
        name: "cropAreas",
        keyName: "key"
});

const TableRow = () => {
        return fields.map((item, index) => {
            return (
                <React.Fragment key={item.key}>
                    <tr>
                        <td colSpan="2" className="custom-td">
                            <label className="checkbox checkbox-primary">
                               //Dont work
                                 <Controller 
                                    name={`cropAreas[${index}].checked`}
                                    as={<input 
                                        type="checkbox"
                                        ref={register()}
                                    />}
                                    control={control}
                                    defaultValue={item.checked}
                                />
                                //Dont work too
                               <input 
                                    name={`cropAreas[${index}].checked`}
                                    type="checkbox"
                                    defaultChecked={item.checked}
                                    ref={register()}
                                />
                                <span>{item.name}</span>
                                <span className="checkmark"></span>
                            </label>
                        </td>
                    </tr>
                </React.Fragment>
            );
        });
    };
```

I pass the defaultValues its ok, but when i click with mouse they stay freeze.

**Comment by bluebill1049** [comment]
no worries, Controller's name have be unique, i think you want to build a wrapper and let the Controller to collect value.

**Reply by GabrielGomesBarbosa** [reply]
Sorry i dont understand, i need to create a new component ? and use here ?

**Reply by bluebill1049** [reply]
<img width="319" alt="Screen Shot 2020-07-31 at 9 18 14 am" src="https://user-images.githubusercontent.com/10513364/88983585-c5d07f80-d30e-11ea-8225-bbc8fa8c4218.png">

see both use the same `name`, Controller's **name** need to be unique.

**Reply by GabrielGomesBarbosa** [reply]
Oh... really sorry @bluebill1049 , im not using the same name... they are with the same name just to show you... how i tried to do... first i tried using without `<Controller/>` just like this: 

```
<input 
   name={`cropAreas[${index}].checked`}
   type="checkbox"
   defaultChecked={item.checked}
   ref={register()}
/>
```

So... its dont work, because the input stay freeze, after (i remove that code) and then i tried like this:

```
<Controller 
  name={`cropAreas[${index}].checked`}
  as={<input 
           type="checkbox"
            ref={register()}
 />}
control={control}
 defaultValue={item.checked}
 />
```

Really sorry for the mistake

**Reply by bluebill1049** [reply]
can you share a CSB?

**Reply by GabrielGomesBarbosa** [reply]
Sorry @bluebill1049 i made it https://codesandbox.io/s/cocky-austin-71s66?file=/src/App.js but its working fine, and my code in my project work well without `Controller` i dont know whats happens... so thanks for everything

**Reply by bluebill1049** [reply]
no worries @GabrielGomesBarbosa :)

**Comment by GabrielGomesBarbosa** [comment, marked as answer]
Hi @bluebill1049 sorry for back here... i still have some questions...

1. In my code in sandbox https://codesandbox.io/s/cranky-goodall-q3cf3?file=/src/App.js when you click in "send" button, in console you will see something like this:

![image](https://user-images.githubusercontent.com/37483241/89427273-7379cf00-d711-11ea-966d-5178ccf2cead.png)

consider that i pass this object in default values:

![image](https://user-images.githubusercontent.com/37483241/89427381-90160700-d711-11ea-82da-b73b7351623e.png)

Its fine, but in my application when i hit the save button, look on that... 

![image](https://user-images.githubusercontent.com/37483241/89427588-c489c300-d711-11ea-9138-0e99e6193271.png)

I receive a object some strange... and my code to defaultValues and my code on component is the same that sandbox (there is more properties, but contains the same ideia that sandbox)

Default Values:
![image](https://user-images.githubusercontent.com/37483241/89429115-85f50800-d713-11ea-8d05-2cd686dd2fb8.png)

My Code (Its the same the sandbox):
![image](https://user-images.githubusercontent.com/37483241/89428314-a40e3880-d712-11ea-9140-9aad4b822203.png)

When i hit in save button:
![image](https://user-images.githubusercontent.com/37483241/89428232-8d67e180-d712-11ea-82b5-80fce8a2134b.png)

Did you understand ? Whats is wrong ?

2. When i hit the send button, some properties dont appear on `console.log()` like `name` propertie, my question is, to get these aditional propertie i need to use input hidden passing the item ? like this ?

`<input type="hidden" value={'propertie[${index}].name'} ref={register()}>`

**Reply by bluebill1049** [reply]
are you using the latest version in your project?

**Reply by GabrielGomesBarbosa** [reply]
I'm using `5.7.2`.

**Reply by bluebill1049** [reply]
hmmmm.... strange, it should work, why in your example result is `array`....

**Reply by bluebill1049** [reply]
any chance you can reproduce with your app in a codesandbox?

**Reply by GabrielGomesBarbosa** [reply]
I dont know exactly what is wrong... because the code is the same that documentation... i will try again! i will build a new sandbox copying part of my real code, and test if i can reproduce! Thanks for while... But the code is correct right ? 😂 For a instant.. I think that i was crazy or making some mistake...

**Reply by bluebill1049** [reply]
yea... the code looks correct. it's a strange one, but must be something wrong. I will take a look again when you got the CSB.

**Reply by bluebill1049** [reply]
wait one thing, why is const TableRow not it's own component? you know this will mount and unmount component each render right?

**Reply by GabrielGomesBarbosa** [reply]
hummm... good point, i dind't create a new component for `<TableRow>` because when i use the `useFieldArray` in simple component. For example... A simple form that have a  `name` and list of checkbox like `listPermition` the list stay intact and fields works... So... i was guess that when i create others components, the re-render will not affect that.

**Reply by GabrielGomesBarbosa** [reply]
But... i'm not sure, i will test it... Thank's @bluebill1049 i will back with result

**Reply by GabrielGomesBarbosa** [reply]
Hi @bluebill1049 you was right now its working i will describe the changes:

```
const MyGlobalComponent = () => {
      
      //react-hook-form methods
      const formMethods = useForm();
      const array = useFieldArray(....);

      const ListCheckBox = () => { ... All Checkbox component } 

     return ( 
       <MyBodyComponent>
           ...
           <ListCheckBox />
            ...
       </MyBodyComponent>
      )
}
```

I change this above to bellow:

```
const MyGlobalComponent = () => {
      
      //react-hook-form methods
      const formMethods = useForm();
      const array = useFieldArray(....);

     return ( 
       <MyBodyComponent>
           ...
           { ... All Checkbox component } 
            ...
       </MyBodyComponent>
      )
}
```

And now its working fine... I just have 1 more question, do you have some example to pass and retrieve more data from `useFieldArray` ? because the final obeject is just the `check` propertie to get name, id... etc. I put many `<input type="hidden" name="propertie[index].prop[index].name" />`  to get all object properties on save method.

Thanks!

**Reply by bluebill1049** [reply]
I don't have more examples, but if you need many input hidden, probably worth considering a data merge at final submit.

**Reply by GabrielGomesBarbosa** [reply]
OK... dont worry about it, thanks a lot @bluebill1049
